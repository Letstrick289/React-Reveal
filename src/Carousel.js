/*
 * Carousel Component For react-reveal
 *
 * Copyright Â© Roman Nosov 2017
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React from 'react';
import { number, element, node} from 'prop-types';
import Skin from './lib/CarouselSkin';

const
  propTypes = {
    skin: element,
    children: node.isRequired,
    wait: number,
    maxTurns: number,
  },
  defaultProps = {
    skin: <Skin />,
    wait: 5000,
    maxTurns: 2,
  };

class Carousel extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      prev: React.Children.count(this.props.children) - 1,
      next: 0,
      backwards: false,
      swap: false
    };
    this.turn = 0;
    this.handleReveal = this.handleReveal.bind(this);
    this.next = this.next.bind(this);
    this.prev = this.prev.bind(this);
    this.target = this.target.bind(this);
  }

  target({target}) {
    let newPos = +target.getAttribute('data-index');
    this.setState({
      next: newPos,
      prev: this.state.next,
      backwards: newPos<this.state.next,
      swap: !this.state.swap
    });
  }

  handleReveal() {
    if (this.turn>=this.props.maxTurns)
      return;
    this.next();
  }

  move(newPos) {
    let pos = newPos;
    const count = React.Children.count(this.props.children);
    if (newPos >= count) {
      this.turn++
      pos = 0;
    }
    else if (newPos < 0)
      pos = count -1;
    this.setState({
      next: pos,
      prev: this.state.next,
      backwards: newPos<this.state.next,
      swap: !this.state.swap
    });
  }
//  d = Date.now();
  next() {
//    console.log('next', ~~((-this.d + Date.now())/1000));
    this.move(this.state.next + 1);
  }

  prev() {
    this.move(this.state.next - 1);
  }

  render() {
    const children = React.Children.toArray(this.props.children),
          count = children.length;
    if (count<2)
      return <div>{children}</div>;
    let { swap, prev, next, backwards } = this.state;
    next %= count; prev %= count;
    let before = children[swap ? prev : next];
    let after  = children[swap ? next : prev];
    return (
      <this.props.skin.type {...this.props.skin.props} api={{
        position: next,
        handleClick: this.target,
        total: count,
        props: this.props,
        next: this.next,
        prev: this.prev,
      }}>
        <before.type
          wait={this.props.wait}
          {...before.props}
          style={{
            position: 'static',
            left: 0,
            top: 0,
          }}
          opposite
          when={!swap}
          mirror={backwards}
          onReveal={!swap ? this.handleReveal : void 0}
        />
        <after.type
          wait={this.props.wait}
          {...after.props}
          style={{
            position: 'absolute',
            left: 0,
            top: 0,
          }}
          opposite
          when={swap}
          mirror={backwards}
          onReveal={swap ? this.handleReveal : void 0}
        />
      </this.props.skin.type>
    );
  }

}

Carousel.propTypes = propTypes;
Carousel.defaultProps = defaultProps;
export default Carousel;
