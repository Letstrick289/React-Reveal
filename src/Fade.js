/*
 * Fade React Component
 *
 * Copyright Â© Roman Nosov 2017
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React from 'react';
import { bool } from 'prop-types';
import Base from './Base';
import { animation } from './lib/globals';

const
  propTypes = {
    left: bool,
    right: bool,
    top: bool,
    bottom: bool,
    big: bool,
  },
  defaultProps = {
    
  };

function Fade({ left, right, up, down, top, bottom, big, ...props }) {
	const 
		dist = big ? '2000px' : '100%', 
		effect=`from { opacity: 0;${left||right||up||down||top||bottom ? ` transform: translate3d(${left?`-${dist}`:(right?dist:'0')}, ${down||top?`-${dist}`:(up||bottom?dist:'0')}, 0);` : ''}}`;
  return <Base {...props} animation={animation(effect)} />;
}


Fade.propTypes = propTypes;
Fade.defaultProps = defaultProps;
export default Fade;