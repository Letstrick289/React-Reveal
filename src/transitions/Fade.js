/*
 * Fade React Component
 *
 * Copyright Â© Roman Nosov 2017
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React from 'react';
import { bool } from 'prop-types';
import Reveal from '../Reveal';
import { animation } from '../lib/globals';

const
  propTypes = {
    out: bool,
    left: bool,
    right: bool,
    top: bool,
    bottom: bool,
    big: bool,
  },
  defaultProps = { 

  };

function Fade({out, left, right, up, down, top, bottom, big, ...props}, context) {
  
  function factory(reverse) {
    
    function make() {
      const 
        xdist = big ? '2000px' : '100%', ydist =  big ? '2000px' : props.collapse || '100%'; 
        return animation(
          `${!reverse?'from':'to'} {${props.collapse?`max-height:0;`:''}opacity: 0;${left||right||up||down||top||bottom ? ` transform: translate3d(${left?`-${xdist}`:(right?xdist:'0')}, ${down||top?`-${ydist}`:(up||bottom?ydist:'0')}, 0);` : ''}}
            ${reverse?'from':'to'} {${props.collapse?`max-height:${props.collapse};`:''}opacity: 1;transform: none;} `);  
    }
    
    return { reverse: left, make }; 
  }
  
  return context
    ? <Reveal {...props} in={factory(false)} out={factory(true)} />
    : factory(out)
  ;
}

Fade.propTypes = propTypes;
Fade.defaultProps = defaultProps;
export default Fade;