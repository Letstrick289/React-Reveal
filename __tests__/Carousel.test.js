/*
 * Carousel Test Suite
 *
 * Copyright Â© Roman Nosov 2017
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */
import Carousel from '../Carousel';
import Fade from '../Fade';

import React from 'react';
import { shallow } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';

configure({ adapter: new Adapter() });

describe('Carousel', () => {
  it('renders a initial view', () => {
    const content = shallow(
      <Carousel>
        <Fade>Test test</Fade>
        <Fade>Test test</Fade>
      </Carousel>
    );
    expect(content.html()).toMatchSnapshot();
  });
});
