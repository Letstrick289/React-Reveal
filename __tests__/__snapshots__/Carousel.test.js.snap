exports[`Carousel renders a initial view 1`] = `"<div style=\"position:relative;width:500px;height:250px;overflow:hidden;\"><div style=\"position:relative;height:200px;\"><div class=\"react-reveal\" style=\"position:static;left:0;top:0;\">Test test</div><div class=\"react-reveal\" style=\"position:absolute;left:0;top:0;opacity:0;\">Test test</div></div><div style=\"line-height:250px;text-align:center;position:absolute;left:0;top:0;width:10%;font-size:4em;cursor:pointer;user-select:none;\"> &lt; </div><div style=\"line-height:250px;text-align:center;position:absolute;left:90%;top:0;width:10%;font-size:4em;cursor:pointer;\"> &gt; </div><div style=\"text-align:center;\"><span style=\"font-size:1.5em;cursor:pointer;\" data-index=\"0\"> ● </span><span style=\"font-size:1.5em;cursor:pointer;\" data-index=\"1\"> ○ </span></div></div>"`;
